mempool/src/shared_mempool/types.rs

```diff
@@ -305,33 +333,75 @@ impl From<Vec<u64>> for MultiBucketTimelineIndexIds {
     }
 }
 
+// MempoolMessageId is used to identify a mempool message (set of transactions) sent to a peer.
+// Each tuple (a, b) indicates a range of the timeline index IDs read from the core mempool
+// timeline index that produced the txns in this message.
+// Each tuple (a, b) is encoded as follows:
+// The first byte of a, b indicates the sender bucket (hash of the sender address).
+// Each sender bucket is further subdivided into timelines (One timeline per range of txn fees).
+// A timeline is a queue of transactions.
+// The second byte of a, b indicates the index of the timeline inside the sender bucket.
+// The remaining bytes of a, b indicate the start and end pointers inside the timeline.
 #[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq, Serialize)]
-pub struct MultiBatchId(pub Vec<(u64, u64)>);
+pub struct MempoolMessageId(pub Vec<(u64, u64)>);
 
-impl MultiBatchId {
+impl MempoolMessageId {
     pub(crate) fn from_timeline_ids(
-        old: &MultiBucketTimelineIndexIds,
-        new: &MultiBucketTimelineIndexIds,
+        timeline_ids: Vec<(
+            MempoolSenderBucket,
+            (MultiBucketTimelineIndexIds, MultiBucketTimelineIndexIds),
+        )>,
     ) -> Self {
         Self(
-            old.id_per_bucket
+            timeline_ids
                 .iter()
-                .cloned()
-                .zip(new.id_per_bucket.iter().cloned())
+                .flat_map(|(sender_bucket, (old, new))| {
+                    old.id_per_bucket
+                        .iter()
+                        .cloned()
+                        .zip(new.id_per_bucket.iter().cloned())
+                        .enumerate()
+                        .map(move |(index, (old, new))| {
+                            let sender_bucket = *sender_bucket as u64;
+                            let timeline_index_identifier = index as u64;
+                            (
+                                sender_bucket << 56 | timeline_index_identifier << 48 | old,
+                                sender_bucket << 56 | timeline_index_identifier << 48 | new,
```

Comment:
should you assert that sender_bucket and timeline_index_identifier are below 128?

Reply:
Done.